File: core/src/Core/Trading.hs
module Core.Trading where

import Domain.Player
import Domain.GameState
import Domain.Resource
import Core.Resources

import Data.List (sort, nub, (\\))

type TradingOption = [(Domain.Player.Name, ResourceType, Coin)]
type TradeSetup = (Domain.Player.Name, [ResourceProduced], [ResourceType -> Int])

tryTrading :: GameState -> Domain.Player.Name -> [[ResourceCost]] -> [TradingOption]
tryTrading gs@GameState {playerStates = playerStates} playerName =
    computeTradingOptions (
        (neighbourWest, resourcesProducedWest, tradeFunctionsWest),
        (neighbourEast, resourcesProducedEast, tradeFunctionsEast)
    )
    where
        tradeFunctions :: [Domain.Player.Name -> ResourceType -> Int]
        tradeFunctions = extractActions gs $ tradeActions $ lookupPlayerState gs playerName
        (neighbourWest, neighbourEast) = lookupNeighbours gs playerName
        resourcesProducedWest = resourcesProduced $ lookupPlayerState gs neighbourWest
        resourcesProducedEast = resourcesProduced $ lookupPlayerState gs neighbourEast
        tradeFunctionsWest = (\tradeFunction -> tradeFunction neighbourWest) <$> tradeFunctions
        tradeFunctionsEast = (\tradeFunction -> tradeFunction neighbourEast) <$> tradeFunctions


computeTradingOptions :: (TradeSetup, TradeSetup) -> [[ResourceCost]] -> [TradingOption]
computeTradingOptions tradeSetups resourceCostCombinations =
    resourceCostCombinations >>= doComputeTradingOptions tradeSetups
    where
        doComputeTradingOptions :: (TradeSetup, TradeSetup) -> [ResourceCost] -> [TradingOption]
        doComputeTradingOptions (tradeSetupWest, tradeSetupEast) resourceCosts =
            let
                order1 = (tradeSetupWest, tradeSetupEast)
                order2 = (tradeSetupEast, tradeSetupWest)
            in
                nub $ applyTradeSetups order1 resourceCosts ++ applyTradeSetups order2 resourceCosts

applyTradeSetups :: (TradeSetup, TradeSetup) -> [ResourceCost] -> [TradingOption]
applyTradeSetups ((name1, resources1, trade1), (name2, resources2, trade2)) resourceCosts =
    case coverResourceCosts resourceCosts resources1 of
        Nothing -> 
            [computeTradeCost name1 trade1 resourceCosts]
        Just resourceCostCombinationsAfterTrade ->
            resourceCostCombinationsAfterTrade >>= applySecondTradeSetup
            where 
                applySecondTradeSetup remainingResourceCosts = 
                    let
                        currentCost = computeTradeCost name1 trade1 (resourceCosts `diffResourceCosts` remainingResourceCosts)
                    in
                        case coverResourceCosts remainingResourceCosts resources2 of
                            Nothing ->
                                [sort $ currentCost ++ computeTradeCost name2 trade2 remainingResourceCosts]
                            Just _ ->
                                []
        
computeTradeCost :: Domain.Player.Name -> [ResourceType -> Int] -> [ResourceCost] -> TradingOption
computeTradeCost name tradeFunctions resourceCosts =
    sort $ resourceCosts >>= computeResourceCost
    where
        computeResourceCost (Cost resourceType count) = replicate count (name, resourceType, tradeValue resourceType)
        tradeValue resourceType = minimum $ (\tradeFunction -> tradeFunction resourceType) <$> tradeFunctions

diffResourceCosts = diff []
    where
        diff accResourceCosts [] _ = accResourceCosts
        diff accResourceCosts resourceCosts [] = accResourceCosts ++ resourceCosts 
        diff accResourceCosts (resourceCost:resourceCosts) resourceCosts2 =
            diff (doDiff accResourceCosts resourceCost resourceCosts2) resourceCosts resourceCosts2
            where
                doDiff accResourceCosts cost1 [] = accResourceCosts ++ [cost1]
                doDiff accResourceCosts cost1@(Cost resourceType1 c1) (Cost resourceType2 c2 : resourceCosts)
                    | resourceType1 == resourceType2 && c1 <= c2 = accResourceCosts 
                    | resourceType1 == resourceType2 && c1 > c2 = accResourceCosts ++ [Cost resourceType1 (c1 - c2)]
                    | otherwise = doDiff accResourceCosts cost1 resourceCosts

File: core/src/core/trading.rs
use crate::domain::{
    diff_resource_costs, Coin, GameState, PName, PlayerState, ResourceCost, ResourceCostOption,
    ResourceCostOptions, ResourceType,
};
use maplit::hashset;
use std::collections::HashSet;

pub type TradingOptions = HashSet<TradingOption>;
pub type TradingOption = Vec<(PName, ResourceType, Coin)>;

pub fn try_trading(
    game_state: &GameState,
    player: &PName,
    resource_cost_options: ResourceCostOptions,
) -> TradingOptions {
    let current = game_state.get_player_state(player);
    let (west, east) = game_state.get_neighbour_player_states(player);
    compute_trading_options((current, west, east), resource_cost_options)
}

type TradingSetup<'a> = (&'a PlayerState, &'a PlayerState, &'a PlayerState);

fn compute_trading_options(
    (current, west, east): TradingSetup,
    resource_cost_options: ResourceCostOptions,
) -> TradingOptions {
    let mut trading_options: TradingOptions = Default::default();
    for resource_cost_option in resource_cost_options {
        trading_options.extend(apply_trading_setup(
            (current, west, east),
            resource_cost_option.clone(),
        ));
        trading_options.extend(apply_trading_setup(
            (current, east, west),
            resource_cost_option.clone(),
        ));
    }
    trading_options
}

fn apply_trading_setup(
    (current, player_state1, player_state2): TradingSetup,
    resource_costs: ResourceCostOption,
) -> TradingOptions {
    let resource_cost_options1 = player_state1.cover_resource_costs(resource_costs.clone());
    if resource_cost_options1.len() == 0 {
        return hashset![build_trade_option(
            current,
            player_state1,
            resource_costs.clone()
        )];
    }
    let mut trading_options: TradingOptions = Default::default();
    for resource_cost_option in resource_cost_options1 {
        let mut trading_option = build_trade_option(
            current,
            player_state1,
            diff_resource_costs(resource_costs.clone(), resource_cost_option.clone()),
        );
        let resource_cost_options2 =
            player_state2.cover_resource_costs(resource_cost_option.clone());
        if resource_cost_options2.len() == 0 {
            trading_option.extend(build_trade_option(
                current,
                player_state2,
                resource_cost_option.clone(),
            ));
            trading_option.sort();
            trading_options.insert(trading_option);
        }
    }
    trading_options
}

fn build_trade_option(
    current: &PlayerState,
    player_to_trade: &PlayerState,
    resource_costs: Vec<ResourceCost>,
) -> TradingOption {
    let mut trading_option: TradingOption = Default::default();
    let player_name_to_trade = player_to_trade.player.name();
    for ResourceCost(resource_type, resource_count) in resource_costs {
        let trading_value = current.apply_trading(player_name_to_trade, &resource_type);
        for _ in 0..resource_count {
            trading_option.push((player_name_to_trade.clone(), resource_type, trading_value))
        }
    }
    trading_option
}
